src/order/category/omega_complete_partial_order.lean

```diff
@@ -47,4 +58,316 @@ instance : inhabited ωCPO := ⟨of punit⟩
 
 instance (α : ωCPO) : omega_complete_partial_order α := α.str
 
+open category_theory.limits
+
+instance : has_terminal ωCPO.{u} :=
+{ has_limit := λ F,
+  { exists_limit :=
+    ⟨ { cone :=
+        { X := of punit,
+          π := { app := λ X, pempty.elim X } },
+        is_limit :=
+        { lift := λ s, ⟨λ x, punit.star,λ x y h, le_refl _,λ c, punit_eq _ _⟩ } } ⟩ } }
+
+open omega_complete_partial_order category_theory category_theory.limits
+
+/--
+(internal implementation) the limit cone of the binary product in a ωCPO.
+It is implemented as the product type -/
+def product_cone (X Y : ωCPO.{u}) : cone (pair X Y) :=
+binary_fan.mk
+  (continuous_hom.of_mono preorder_hom.prod.fst (λ c, rfl) : ωCPO.of (X × Y) ⟶ _)
+  (continuous_hom.of_mono preorder_hom.prod.snd (λ c, rfl))
+
+/-- (internal implementation) the fact that the proposed product cone is the limit -/
+def product_cone_is_limit (X Y : ωCPO.{u}) : is_limit (product_cone X Y) :=
+{ lift := λ s, ⟨λ x, (s.π.app walking_pair.left x, s.π.app walking_pair.right x),
+               λ x y h, ⟨(s.π.app walking_pair.left).monotone h, (s.π.app walking_pair.right).monotone h⟩,
+               λ c, by ext; dsimp; rw continuous_hom.continuous; refl⟩,
+  fac' := by rintros s ⟨ ⟩; ext; refl,
+  uniq' := by { dsimp, intros,
+                ext; dsimp; delta binary_fan.fst binary_fan.snd; rw ← w;
+                  simp only [continuous_hom.coe_fn_mk, binary_fan.π_app_left, binary_fan.π_app_right];
+                  refl, } }
+
+instance {X Y : ωCPO} : has_limit (pair X Y) :=
+has_limit.mk ⟨_, product_cone_is_limit X Y⟩
+
+instance : has_binary_products ωCPO.{u} :=
+has_binary_products_of_has_limit_pair _
+
+/-- Constructor for values in binary products of ωCPOs, as an arrow from a unit type -/
+noncomputable def prod_lift {X Y : ωCPO.{u}} (x : X) (y : Y) : ωCPO.of punit.{u + 1} ⟶ X ⨯ Y :=
+limits.prod.lift (continuous_hom.const x) (continuous_hom.const y)
+
+/-- Constructor for values in binary products of ωCPOs -/
+noncomputable def prod.mk {X Y : ωCPO.{u}} (x : X) (y : Y) : ↥(X ⨯ Y) :=
+prod_lift x y punit.star
+
+/-- Isomorphism between binary products of ωCPOs and product types -/
+noncomputable def of_prod_iso (X Y : ωCPO.{u}) : X ⨯ Y ≅ ωCPO.of (X × Y) :=
+limits.is_limit.cone_point_unique_up_to_iso (limit.is_limit _) (product_cone_is_limit X Y)
```

Comment:
This could presumably be factored into an instance of an appropriate `preserves_limits_of_shape` class, and then there is a canonical function to obtain the isomorphism. Unfortunately I think this is mid-refactor at the moment, in #4163, by @b-mehta.

Reply:
I don't mind if this goes ahead and I can fix anything that breaks in my refactor!